minicriu-client.c

```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
 	return bytes;
 }
 
+static unsigned long align_up(unsigned long v, unsigned p) {
+	return (v + p - 1) & ~(p - 1);
+}
+
+static int mc_save_core_file() {
+
+	pid_t pid = syscall(SYS_getpid);
+	int phnum = 0;
+
+	// Create Elf header
+	memcpy(ehdr.e_ident, ELFMAG, SELFMAG);
+	ehdr.e_ident[EI_CLASS] = ELFCLASS64;
+	#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+	ehdr.e_ident[EI_DATA] = ELFDATA2LSB;
+	#else
+	ehdr.e_ident[EI_DATA] = ELFDATA2MSB;
+	#endif
+	ehdr.e_ident[EI_VERSION] = EV_CURRENT;
+	ehdr.e_ident[EI_OSABI] = ELFOSABI_SYSV;
+	ehdr.e_type = ET_CORE;
+	ehdr.e_machine = EM_X86_64;
+	ehdr.e_version = EV_CURRENT;
+	ehdr.e_entry = 0;
+	ehdr.e_phoff = sizeof(Elf64_Ehdr);
+	ehdr.e_shoff = 0;
+	ehdr.e_flags = 0;
+	ehdr.e_ehsize = sizeof(Elf64_Ehdr);
+	ehdr.e_phentsize = sizeof(Elf64_Phdr);
+	ehdr.e_phnum = 0;
+	ehdr.e_shentsize = 0;
+	ehdr.e_shnum = 0;
+	ehdr.e_shstrndx = 0;
+
+	// Create PT_NOTE phdr
+	phdr[phnum].p_type = PT_NOTE;
+	phdr[phnum].p_flags = 0;
+	phdr[phnum].p_offset = 0;
+	phdr[phnum].p_vaddr = 0;
+	phdr[phnum].p_paddr = 0;
+	phdr[phnum].p_memsz = 0;
+	phdr[phnum].p_filesz = 0;
+	phdr[phnum++].p_align = 0;
+
+	FILE *proc_maps = fopen("/proc/self/maps", "r");
+	if (proc_maps == NULL) {
+		perror("Could not open maps file. Failed to create checkpoint.");
+		return 1;
+	}
+
+	// Initialize NT_FILE
+	nt_file.descsz = 0;
+	nt_file.descsz += sizeof(nt_file.count) + sizeof(nt_file.page_size);
+	nt_file.page_size = 0x1000;
+
+	// Create PT_LOAD and NT_FILE phdrs
+	char buffer[256];
+	while (fgets(buffer, sizeof(buffer), proc_maps)) {
+		void *addr_start, *addr_end;
+		char perms[8];
+		long ofs;
+		int name_start = 0;
+		int name_end = 0;
+
+		int res = sscanf(buffer, "%p-%p %7s %lx %*d:%*d %*lx %n%*[^\n]%n", &addr_start,
+			&addr_end, perms, &ofs, &name_start, &name_end);
+
+		if (res < 4) {
+			perror("sscanf. Failed to create checkpoint.");
+			fclose(proc_maps);
+			return 1;
+		}
+
+		// [vsyscall] is mapped to the same address in each process
+		if (!strncmp(buffer + name_start, "[vsyscall]", sizeof("[vsyscall]") - 1)) {
```

Comment:
Sorry for such a long reply, I was busy with my studies. To tell you the truth, I'm not sure if `[vvar]` and `[vdso]` should be done the same way.  I don't save `[vsyscall]` for two reasons. First, the process has no rights to read the region and second, in every process the `[vsyscall]` region is mapped to the same address space. As far as I know, the same cannot be said for `[vvar]` and `[vsdo]`.

Moreover, new regions `[vvar]` and `[vdso]` are deleted during checkpoint recovery and old ones are restored in #4. This did not cause any problems on a simple example.

This is an important question, and I am investigating it.

Link to test example -- https://github.com/i-pankrat/crac/commit/a85581f5d0067f3bde07cb1c359a943f3b29d902?diff=unified

Reply:
Sure, that aspect can be handled elsewhere.

In my case I was getting SIGSEGV (I think) when the code tried to read from `[vvar]`, despite that region being mapped with read privileges. And once the JVM is restored, it would try to read current timestamps; accessing the old addresses (in new process where vvar and vdso are on a different memory location) would cause another segfault. Luckily, `mremap` from the new addresses to the old ones was sufficient to fix this.